src/dsa/rpo_falcon512/keys/public_key.rs

```diff
@@ -0,0 +1,135 @@
+        Self(pub_key)
+    }
+
+    /// Verifies the provided signature against provided message and this public key.
+    pub fn verify(&self, message: Word, signature: &Signature) -> bool {
+        signature.verify(message, self.0)
+    }
+}
+
+impl From<PubKeyPoly> for PublicKey {
+    fn from(pk_poly: PubKeyPoly) -> Self {
+        let pk_felts: Polynomial<Felt> = pk_poly.0.into();
+        let pk_digest = Rpo256::hash_elements(&pk_felts.coefficients).into();
+        Self(pk_digest)
+    }
+}
+
+impl From<PublicKey> for Word {
+    fn from(key: PublicKey) -> Self {
+        key.0
+    }
+}
+
+// PUBLIC KEY POLYNOMIAL
+// ================================================================================================
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct PubKeyPoly(pub Polynomial<FalconFelt>);
+
+impl Deref for PubKeyPoly {
+    type Target = Polynomial<FalconFelt>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl From<Polynomial<FalconFelt>> for PubKeyPoly {
+    fn from(pk_poly: Polynomial<FalconFelt>) -> Self {
+        Self(pk_poly)
+    }
+}
+
+impl Serializable for &PubKeyPoly {
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        let mut buf = [0_u8; PK_LEN];
+        buf[0] = 9;
+
+        let mut acc = 0_u32;
+        let mut acc_len = 0;
+
+        let mut input_pos = 1;
+        for c in self.0.coefficients.iter() {
+            let c = c.value();
+            acc = (acc << 14) | c as u32;
+            acc_len += 14;
+            while acc_len >= 8 {
+                acc_len -= 8;
+                buf[input_pos] = (acc >> acc_len) as u8;
+                input_pos += 1;
+            }
+        }
+        if acc_len > 0 {
+            buf[input_pos] = (acc >> (8 - acc_len)) as u8;
+        }
+
+        target.write(buf);
+    }
+}
+
+impl Deserializable for PubKeyPoly {
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        let buf = source.read_array::<PK_LEN>()?;
+
+        if buf[0] != LOG_N as u8 {
+            return Err(DeserializationError::InvalidValue(format!(
+                "Failed to decode public key: expected the first byte to be {LOG_N} but was {}",
+                buf[0]
+            )));
+        }
+
+        let mut acc = 0_u32;
+        let mut acc_len = 0;
+
+        let mut output = [FalconFelt::zero(); N];
+        let mut output_idx = 0;
+
+        for &byte in buf.iter().skip(1) {
+            acc = (acc << 8) | (byte as u32);
+            acc_len += 8;
+
+            if acc_len >= 14 {
+                acc_len -= 14;
+                let w = (acc >> acc_len) & 0x3FFF;
+                if w >= MODULUS as u32 {
+                    return Err(DeserializationError::InvalidValue(format!("Failed to decode public key: coefficient {w} is greater than or equal to the field modulus {MODULUS}")));
+                }
+                output[output_idx] = FalconFelt::new((w) as i16);
```

Comment:
We should probably implement `TryFrom<u32> for FalconFelt` and handle error checking there rather than here.

Reply:
This is more related to a wrong encoding so I think it is more relevant here. I can also move it if you feel strongly about it.

Reply:
I think there are two parts to this: (1) the value is outside of the valid range for the field element and (2) public key is encoded incorrectly. I would still probably have the `TryFrom<u32 for FalconFelt` and this would check the first part (the error returned could be just a string). Then, this error could be caught here, and we can pre-pend the "Failed to decode public key" part to it.

Reply:
changed it now